### PR TITLE
Switch to single, mandatory PartitionKey

### LINKED ISSUE #155: CosmosDB: make PartitionKey mandatory
While it was possible to use certain SKUs in the past that allowed a paritionless container this is no longer the case.
We should follow through and make it mandatory in the SDK as well.

For example [create document](https://github.com/Azure/azure-sdk-for-rust/blob/master/sdk/cosmos/src/requests/create_document_builder.rs) makes partition key(s) optional and should be changed. 

I also would like to take the time to discuss a partition key reimagining. Right now it's treated as an independent field from the serialized type while it's clearly not the case. 

For example, let's take the code of [examples/document_00.rs](https://github.com/Azure/azure-sdk-for-rust/blob/master/sdk/cosmos/examples/document_00.rs). 

We have a document defined like this:

[https://github.com/Azure/azure-sdk-for-rust/blob/bee90032290089297cfdd0c9f7ada3518890e2cb/sdk/cosmos/examples/document_00.rs#L13-L19](https://github.com/Azure/azure-sdk-for-rust/blob/bee90032290089297cfdd0c9f7ada3518890e2cb/sdk/cosmos/examples/document_00.rs#L13-L19)

Right now the only constraints in order to use it in Cosmos are:

1. to be `Serialize` and `Deserialize` (in practice, they are not both required at the same time  if you just want to insert/retrieve).
2. Have a field that matches the partition key specified in the container creation.

1 is enforced, 2 is not. 

What makes thing worse is that, right now, in every `CreateDocument`, `ReplaceDocument` etc... we accept an *arbitrary* value.

For example:

[https://github.com/Azure/azure-sdk-for-rust/blob/bee90032290089297cfdd0c9f7ada3518890e2cb/sdk/cosmos/examples/document_00.rs#L142-L146](https://github.com/Azure/azure-sdk-for-rust/blob/bee90032290089297cfdd0c9f7ada3518890e2cb/sdk/cosmos/examples/document_00.rs#L142-L146)

This is clearly a poor design solution because it allows the user to pass whatever they want to `partition_keys()` even tough the correct value is already in the serialized struct. Needless to say, a value that does not match the partition key will make the call fail.

In other words, we should be able to extract it automatically without asking for it at all. 

The call would then become:

```rust
 let create_document_response = collection_client 
     .create_document() 
     .execute(&doc) 
     .await?; 
```

Less verbose, and much less error prone.

One approach is to have a trait that must be implemented on every document. Something like:

```rust
pub trait CosmosDocument {
   fn partition_keys(&self) -> &PartitionKeys
}
```

So the code above will be amended like this:

```rust
impl<'a> CosmosDocument  for MySampleStruct<'a> {
  fn partition_keys(&self) -> &PartitionKeys {
      &self.id
  }
}
```

This way we could extract the *correct* value at runtime instead of asking for it from the user. 

This could also be automated via proc macro so the implementation becomes automatic. We could end up with something akin to `serde`:

```rust
 #[derive(Serialize, Deserialize, Debug)] 
 struct MySampleStruct<'a> { 
     #[cosmos(PartitionKey)]
     id: Cow<'a, str>, 
     a_string: Cow<'a, str>, 
     a_number: u64, 
     a_timestamp: i64, 
 } 
```

What do you think?

### PR DESCRIPTION
**Pull request by MindFlavor**
This PR does basically these things:

1. Removes the support for multiple partition keys. While the REST APIs expect a vector, only a vector with a *single* element is supported. The crate now reflects this by accepting a single partition key.
2. Partition Key is now mandatory (wherever applicable). Since CosmosDB has effectively deprecated containers without partition key we change our crate to make it mandatory. It's not mandatory everywhere (for example, queries support the cross partition option) but where it is it's enforced.
3. Intoduces the `CosmosEntity` trait that automates the partition key handling. By implementing this trait (it's just a single one line function) the CosmosDB crate will automatically take care of passing the partition key whenever necessary. This simplifies the API a bit (coupled with 2, it simplifies the API **a lot**). A separate function is available to support the old model, if you so choose (but you have to handle the partition key yourself).

## Example
 
For example suppose we have this struct:

```rust
#[derive(Serialize, Deserialize, Clone, Debug)]
struct MySampleStruct {
    id: String,
    a_string: String,
    a_number: u64,
    a_timestamp: i64,
}
```

We know that the partition key is the `id`. Previously an insert would have been like this:

```rust
  client
        .create_document()
        .partition_keys([&doc.document.id])  // notice the ugly self reference
        .execute(&doc)
        .await
```

Now you can just implement the `CosmosEntity` trait. By doing so, you instruct the crate that the `id` field is your partition key:

```rust
impl<'a> azure_cosmos::CosmosEntity<'a, &'a str> for MySampleStruct {
    fn partition_key(&'a self) -> &'a str {
        self.id.as_ref() // we just return our id field (not even the same type, just a ref is enough)
    }
}
```

Now the insert is just:

```rust
client.create_document().execute(&doc).await  // notice we no longer need to specify the partition key!
```

--- 

While now is manual, it's possible to automate the implementation but using a custom derive (it's just an `as_ref()` call after all).

### Old model

If you want to use the old model - because for example you cannot implement the `CosmosEntiry` trait since you do not define the struct - you can just call `.execute_with_partition_key(<document>, <partition_key>)` as before (notice that both the document and the partition key must still be `Serializable`).

Fixes #155, removes some dependency from `ToJsonVector` (issue #104).

### CODE REVIEW

File: sdk/cosmos/src/requests/create_document_builder.rs [hunk 1]
```diff
@@ -49,15 +51,19 @@ impl<'a, 'b> CreateDocumentBuilder<'a, 'b> {
...
         is_upsert: bool => if is_upsert { IsUpsert::Yes } else { IsUpsert::No },
         indexing_directive: IndexingDirective,
-        partition_keys: PartitionKeys => Some(partition_keys),
     }
 }
 
-impl<'a, 'b> CreateDocumentBuilder<'a, 'b> {
-    pub async fn execute<T: Serialize>(
+impl<'a, 'b, 'c> CreateDocumentBuilder<'a, 'b> {
+    async fn execute_internal<DOC, FNPK>(
```
Comment (rylev) [inline sdk/cosmos/src/requests/create_document_builder.rs:58]: Nit: maybe call it `perform_execute`?

**Review by rylev — CHANGES_REQUESTED** [review]
Looks good. Mostly small nits, but question about the `CosmosEntity` trait.

File: sdk/cosmos/src/cosmos_entity.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+use crate::headers;
+use http::request::Builder;
+use serde::Serialize;
+
+/// CosmosDB partition key. Every CosmosDB entity must implement it.
+pub trait CosmosEntity<'a, T: Serialize + 'a> {
```
[the remark this answers is not in the document]
Comment (MindFlavor) [reply sdk/cosmos/src/cosmos_entity.rs:6]: No reason aside personal preference. I've changed to an AT and the type signature is indeed simpler. For example `create_document_builder::execute` goes from:

```rust
pub async fn execute<PK: Serialize + 'c, T: Serialize + CosmosEntity<'c, PK>>
```
 to 

```rust
pub async fn execute<T: Serialize + CosmosEntity<'c>>
```

File: sdk/cosmos/src/requests/create_document_builder.rs (hunk 1 above)
Comment (MindFlavor) [reply sdk/cosmos/src/requests/create_document_builder.rs:58]: mmm `perform_execute` seems a bit weird (they are synonyms so it's like saying the same thing twice) but I trust your judgement 👍

**Review by rylev — APPROVED** [review]

**Pull request merged**